text/0000-all-the-clones.md

```diff
@@ -0,0 +1,113 @@
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+Make `clone` a lang-item, add the following trait rules to the compiler:
+
+```
+n number
+T type
+T: Clone
+----------
+[T; n]: Clone
+
+T1,...,Tn types
+T1: Clone, ..., Tn: Clone
+----------
+(T1, ..., Tn): Clone
+```
+
+And add the obvious implementations of `Clone::clone` and `Clone::clone_from` as MIR shim implementations, in the same manner as `drop_in_place`. The implementations could also do a shallow copy if the type ends up being `Copy`.
+
+Remove the macro implementations in libcore. We still have macro implementations for other "derived" traits, such as `PartialEq`, `Hash`, etc.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+The MIR shims add complexity to the compiler. Along with the `derive(Clone)` implementation in `libsyntax`, we have 2 separate sets of implementations of `Clone`. 
+
+Having `Copy` and `Clone` impls for all arrays and tuples, but not `PartialEq` etc. impls, could be confusing to users.
+
+# Rationale and Alternatives
+[alternatives]: #alternatives
+
+Even with all proposed expansions to Rust's type-system, for consistency, the compiler needs to have at least *some* built-in `Clone` implementations: the type `for<'a> fn(Foo<'a>)` is `Copy` for all user-defined types `Foo`, but there is no way to implement `Clone`, which is a supertrait of `Copy`, for it (an `impl<T> Clone for fn(T)` won't match against the higher-ranked type).
+
+The MIR shims for `Clone` of arrays and tuples are actually pretty simple and don't add much complexity after we have `drop_in_place` and shims for `Copy` types.
+
+## The array situation
+
+In Rust 1.19, arrays are `Clone` only if they are `Copy`. This code does not compile:
+```Rust
+fn main() {
+    let x = [Box::new(0)].clone(); //~ ERROR
+    println!("{:?}", x[0]);
+}
+```
+
+~~The reason (I think) is that there is no good way to write a variable-length array expression in macros. This wouldn't be fixed by the first iteration of const generics.~~ Actually, this can be done using a for-loop (`ArrayVec` is used here instead of a manual panic guard for simplicity, but it can be easily implemented given const generics).
+```Rust
+impl<const n: usize; T> Clone for [T; n] {
+    fn clone(&self) -> Self {
+        unsafe {
+            let result : ArrayVec<Self> = ArrayVec::new();
+            for elem in (self as &[T]) {
+                result.push(elem.clone());
+            }
+            result.into_inner().unwrap()
+        }
+    }
+}
+```
+
+OTOH, this means that making non-`Copy` arrays `Clone` is less of a bugfix and more of a new feature. It's however a nice feature - `[Box<u32>; 1]` not being `Clone` is an annoying and seemingly-pointless edge case.
+
+## Implement `Clone` only for `Copy` types
+
+As of Rust 1.19, the compiler *does not* have the `Clone` implementations, which causes ICEs such as rust-lang/rust#25733 because `Clone` is a supertrait of `Copy`.
```

Comment:
For convenience, please make these real links.

Reply:
Make what real links?

Reply:
@arielb1 Change 

```markdown
... ICEs such as rust-lang/rust#25733 because ...
```

to

```markdown
... ICEs such as [rust-lang/rust#25733] because ...

[rust-lang/rust#25733]: https://github.com/rust-lang/rust/issues/25733
```

Similar for references to "RFC #XXXX".